larq_compute_engine/mlir/transforms/optimize.cc

```diff
@@ -71,44 +71,117 @@ DenseElementsAttr Bitpack(PatternRewriter& builder, Attribute x) {
 struct SetBconvReadWriteBitpacked : public OpRewritePattern<TF::LceBconv2dOp> {
   using OpRewritePattern<TF::LceBconv2dOp>::OpRewritePattern;
 
-  LogicalResult matchAndRewrite(TF::LceBconv2dOp bconv_op,
+  LogicalResult matchAndRewrite(TF::LceBconv2dOp outer_bconv_op,
                                 PatternRewriter& rewriter) const override {
-    Value bconv_input = bconv_op.input();
-    if (!bconv_input.hasOneUse()) return failure();
+    /*************
+     * Match ops *
+     *************/
 
-    auto bconv_input_op =
-        dyn_cast_or_null<TF::LceBconv2dOp>(bconv_input.getDefiningOp());
-    if (!bconv_input_op) return failure();
-
-    const auto inner_tensor_type = bconv_input_op.getType().cast<ShapedType>();
-
-    // We use 32-bit bitpacking.
-    constexpr int bitwidth = 32;
-
-    // Don't apply this transformation if the inner tensor type is already I32.
-    if (inner_tensor_type.getElementType().isInteger(bitwidth))
-      return failure();
-
-    if (bconv_input_op.padding() == "SAME" && bconv_input_op.pad_values() != 1)
+    if (!outer_bconv_op.input().hasOneUse()) {
       return failure();
+    }
 
-    const auto inner_tensor_shape = inner_tensor_type.getShape();
-    if (inner_tensor_shape.size() != 4) return failure();
-
-    const auto channels = inner_tensor_shape[3];
-    const auto packed_channels = (channels + bitwidth - 1) / bitwidth;
+    auto intermediate_maxpool_op = dyn_cast_or_null<TFL::MaxPool2DOp>(
+        outer_bconv_op.input().getDefiningOp());
 
-    RankedTensorType new_inner_tensor_type =
-        RankedTensorType::get({inner_tensor_shape[0], inner_tensor_shape[1],
-                               inner_tensor_shape[2], packed_channels},
-                              rewriter.getIntegerType(bitwidth));
-
-    rewriter.replaceOpWithNewOp<TF::LceBconv2dOp>(
-        bconv_input_op, new_inner_tensor_type, bconv_input_op.getOperands(),
-        bconv_input_op.getAttrs());
-    rewriter.replaceOpWithNewOp<TF::LceBconv2dOp>(bconv_op, bconv_op.getType(),
-                                                  bconv_op.getOperands(),
-                                                  bconv_op.getAttrs());
+    if (intermediate_maxpool_op &&
+        (!intermediate_maxpool_op.input().hasOneUse() ||
```

Comment:
If I understand correctly, this part tries to match   `bconv2d( maxpool(x) )`, and here its important that the output of `maxpool(x)` has no other uses. However this line of code tests if `x`, the input of maxpool has one use. Why should that matter?

Reply:
Ah good point, I think this is from previously when we didn't support just `maxpool -> bconv`. I'll fix it.